### ISSUE TITLE
`array.__deepcopy__` does not have the correct call signature

### ISSUE THREAD

**Issue by piyueh** [nv-legate/legate.numpy#19]
## Problem

When doing a deep copy (using Python's native `copy` module), an error is raised due to a missing argument in the call signature of `array.__deepcopy__`:
https://github.com/nv-legate/legate.numpy/blob/2b460c5dfdd60b673e37e25231bf625fdf3ead0e/legate/numpy/array.py#L303-L306

## To reproduce

1. step 1: create `test.py`:
   ```python
   import copy
   from legate import numpy
   a = numpy.arange(100)
   b = copy.deepcopy(a)
   ```
2. step 2: run  with `legate --cpus 1 ./test.py -lg:numpy:test`

## Error traceback:
```python
Traceback (most recent call last):
  File "<prefix>/lib/python3.8/site-packages/legion_top.py", line 394, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "<prefix>lib/python3.8/site-packages/legion_top.py", line 193, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "./test.py", line 4, in <module>
    b = copy.deepcopy(a)
  File "<prefix>/lib/python3.8/copy.py", line 153, in deepcopy
    y = copier(memo)
TypeError: __deepcopy__() takes 1 positional argument but 2 were given
```

## Notes

1. Though it's not shown in the traceback, but once I add a second argument `memo` to the definition of `array.__deepcopy__`, the error is gone. (i.e., changing [line 303 in array.py](https://github.com/nv-legate/legate.numpy/blob/2b460c5dfdd60b673e37e25231bf625fdf3ead0e/legate/numpy/array.py#L303) from `def __deepcopy__(self):` to `def __deepcopy__(self, memo):`) However, I don't know if the result of the copying is correct or not.

2. According to the last paragraph of `copy`'s [documentation](https://docs.python.org/3/library/copy.html), an additional argument (additional to `self`) is required in `__deepcopy__`:
   > In order for a class to define its own copy implementation, it can define special methods \_\_copy\_\_() and \_\_deepcopy\_\_(). ... ... The latter is called to implement the deep copy operation; it is passed one argument, the `memo` dictionary. ...

3. Though `ndarray` has its own `.copy()` method, but the `__copy__` and `__deepcopy__` from native Python is still useful. For example, when using a class:
   ```python
   class DummyMesh:
       def __init__(self, bg, ed, n):
           self.vertices = numpy.linspace(bg, ed, n+1)
   ```
   it's easier to copy an instance using `copy.deepcopy`, e.g., `grid_a = DummyMesh(0., 1., 10); grid_b = copy.deepcopy(grid_a)`. In this situation, the `__deepcopy__` of the `ndarray` is triggered. Otherwise, users have to write more lines of code just to make a deep copy of the instance `grid_a`. 

4. From the definition of the shallow copy `array.__copy__` (line 298) and deep copy `array.__deepcopy__`, it seems they both are doing the same thing. Is this also the default behavior in vanilla NumPy? Just curious about this.

**Comment by lightsighter** [comment]
Fixed with `64ce327f939e5`

**Comment by piyueh** [comment]
I confirm this if fixed in the master branch. Closing the issue.
